File: 2024/Day 18/Python/main.py
"""Advent of Code - 2024 - Day 18"""
from collections import deque

Vec = tuple[int, int]
VecSet = set[Vec]
VecList = list[Vec]
VecQueue = deque[Vec]
DistDict = dict[Vec, int]

def bfs(size: int, corrupted: VecSet) -> (int | None):
    """breadth first search"""
    def neighbors(n: Vec):
        for direction in [(0,1),(1,0),(0,-1),(-1,0)]:
            new = (n[0] + direction[0], n[1] + direction[1])
            if (new not in corrupted) and 0 <= new[0] <= size and 0 <= new[1] <= size:
                yield new

    q: VecQueue = deque()
    q.append((0, 0))
    dist: DistDict = {(0, 0): 0}

    while q:
        current = q.popleft()
        if current == (size, size):
            return dist[current]
        for neighbor in neighbors(current):
            if neighbor not in dist:
                dist[neighbor] = dist[current] + 1
                q.append(neighbor)
    return None

def binary_search(array: VecList, low: int, high: int):
    """binary search"""
    while low <= high:
        mid = (low + high) // 2
        if bfs(SIZE, set(array[:(mid+1)])):
            low = mid + 1
        elif low != mid:
            high = mid
        else:
            return mid
    assert False

#SIZE = 6
#BYTES = 12

SIZE = 70
BYTES = 1024

def parse_line(line: str):
    """parse a line of the input"""
    line_split = list(map(int, line.split(",")))
    return (line_split[0], line_split[1])

with open(0, encoding="utf-8") as f:
    all_corrupted = list(map(parse_line, f.read().splitlines()))
print(bfs(SIZE, set(all_corrupted[:BYTES])))

s = binary_search(all_corrupted, BYTES+1, len(all_corrupted)-1)
print(f"{all_corrupted[s][0]},{all_corrupted[s][1]}")

File: 2024/Day 18/C++/main.cpp
#include <string>
#include <unordered_set>
#include <iostream>
#include <queue>
#include <unordered_map>
#include <limits>
#include <optional>

template <class A, class B>
struct std::hash<std::pair<A, B>>
{
    size_t operator()(const pair<A, B> &p) const
    {
        return ((hash<A>{}(p.first) << 1) | (hash<A>{}(p.first) >> (std::numeric_limits<size_t>::digits - 1))) ^ hash<B>{}(p.second);
    }
};

template <typename T, typename U>
std::pair<T, U> operator+(const std::pair<T, U> &l, const std::pair<T, U> &r)
{
    return {l.first + r.first, l.second + r.second};
}

using Vec = std::pair<int, int>;

const int SIZE = 70;
const int BYTES = 1024;
const Vec dirs[] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

std::vector<Vec> neighbors(std::unordered_set<Vec> corrupted, Vec n)
{
    std::vector<Vec> out;
    for (auto direction : dirs)
    {
        Vec next = n + direction;
        if (corrupted.find(next) == corrupted.end() && 0 <= next.first && next.first <= SIZE && 0 <= next.second && next.second <= SIZE)
        {
            out.push_back(next);
        }
    }
    return out;
}

std::optional<int> bfs(std::unordered_set<Vec> corrupted)
{
    std::queue<Vec> q;
    q.push({0, 0});
    std::unordered_map<Vec, int> dist = {{{0, 0}, 0}};
    while (!q.empty())
    {
        Vec current = q.front();
        q.pop();
        if (current == std::make_pair(SIZE, SIZE))
        {
            return dist[current];
        }
        for (Vec neighbor : neighbors(corrupted, current))
        {
            if (dist.find(neighbor) == dist.end())
            {
                dist[neighbor] = dist[current] + 1;
                q.push(neighbor);
            }
        }
    }
    return {};
}

int binary_search(std::vector<Vec> all_corrupted)
{
    int low = BYTES + 1;
    int high = all_corrupted.size() - 1;
    while (low <= high)
    {
        int mid = (low + high) / 2;
        std::unordered_set<Vec> corrupted(all_corrupted.begin(), all_corrupted.begin() + mid + 1);
        if (bfs(corrupted).has_value())
        {
            low = mid + 1;
        }
        else if (low != mid)
        {
            high = mid;
        }
        else
        {
            return mid;
        }
    }
    return -1;
}

int main(int argc, char *argv[])
{
    std::string line;
    std::vector<Vec> all_corrupted;
    while (std::getline(std::cin, line))
    {
        int com = line.find(",");
        all_corrupted.push_back({std::stoi(line.substr(0, com)), std::stoi(line.substr(com + 1))});
    }
    std::unordered_set<Vec> corrupted(all_corrupted.begin(), all_corrupted.begin() + BYTES);
    std::cout << bfs(corrupted).value() << "\n";
    int s = binary_search(all_corrupted);
    std::cout << all_corrupted[s].first << "," << all_corrupted[s].second << "\n";
}